### PR TITLE
Add Support for RISC-V GCC and 64-bit RISC-V LLVM

### PR DESCRIPTION
**Pull request by palmer-dabbelt**
I thought I'd come back and revisit this, as I wanted to play around with compiler-explorer today and noticed it still doesn't support RISC-V.  There are a handful of existing issues

https://github.com/mattgodbolt/compiler-explorer/issues/334 : Clang support, which has been partially resolved and should be fully resolved by this commit.
https://github.com/mattgodbolt/compiler-explorer/issues/965 : GCC support, which should be fully resolved by this commit.
https://github.com/mattgodbolt/compiler-explorer/pull/964 : An obsolete pull request only containing the GCC support, that I closed because crosstool-ng didn't support RISC-V at that time.
https://github.com/mattgodbolt/compiler-explorer-image/pull/176 : A pull request to add a RISC-V GCC 8.2.0 to the image.

There's a lot of information in there, but the short of it is that I think these three patches (two here, one in -image) should get us reasonable RISC-V support.

### CODE REVIEW

File: etc/config/c++.amazon.properties [hunk 1]
```diff
@@ -241,7 +244,7 @@ compiler.zapcc190308.name=x86-64 Zapcc 190308
 
 ###############################
 # Cross GCC
-group.cross.compilers=&ppc:&mips:&msp:&gccarm:&avr
+group.cross.compilers=&ppc:&mips:&msp:&gccarm:&avr:%riscv
```
Comment (partouf) [inline etc/config/c++.amazon.properties]: % should be &

File: etc/config/c++.amazon.properties [hunk 2]
```diff
@@ -370,6 +373,15 @@ compiler.mips564el.exe=/opt/compiler-explorer/mips64el/gcc-5.4.0/mips64el-unknow
 compiler.mips564el.name=MIPS64 gcc 5.4 (el)
 compiler.mips564el.semver=5.4
 
+###############################
+# GCC for RISC-V
+group.riscv.compilers=riscv820
+group.riscv.groupName=RISC-V GCC
+group.riscv.isSemVer=true
+compiler.riscv820.exe=/opt/compiler/explorer/riscv64/gcc-8.2.0/riscv64-unknown-linux-gnu/bin/riscv64-unknown-linux-gnu-g++
```
Comment (partouf) [inline etc/config/c++.amazon.properties]: compiler-explorer instead of compiler/explorer

**Comment by mattgodbolt** [comment]
Thanks @partouf  - I'm cool with this once Partouf's comments have been addressed :)

**Comment by mattgodbolt** [comment]
(I'll also have to get the builder working, which I'm waiting on now)

**Comment by palmer-dabbelt** [comment]
I think I've fixed @partouf's issues -- sorry about that, I wasn't really paying any attention to this.

**Comment by mattgodbolt** [comment]
Thanks! I'll merge once I have things built :)

**Comment by mattgodbolt** [comment]
Compilation is taking much longer than I expected (for the regular daily built compilers). sorry this may have to wait til tomorrow or the weekend!

**Comment by mattgodbolt** [comment]
It finished! Will merge and deploy to beta! yay!

**Comment by mattgodbolt** [comment]
@palmer-dabbelt this is now live on https://godbolt.org/beta -- can you explain what the difference between riscv32 and riscv64 is? The code generated appears identical, at least if I use this example: https://godbolt.org/beta/z/i1YRyp

my concern is the flags aren't working the way I thught they would?

**Comment by palmer-dabbelt** [comment]
The "riscv32" and "riscv64" names are a bit of a historical accident.  The RISC-V ISA has four base ISAs:

* rv32e: 16 integer registers, each 32 bits wide.
* rv32i: 32 integer registers, each 32 bits wide.
* rv64i: 32 integer registers, each 64 bits wide.
* rv128: 32 integer registers, each 128 bits wide.

When designing the tuples we originally just called everything "riscv" (so, "riscv-unknown-linux-gnu-gcc"), but late in the process someone suggested that we split the 32-bit and 64-bit tuples and as a result we now have "riscv32" and "riscv64".  At the time we didn't have rv32e at all, so it seemed like a reasonable choice, but with the addition of rv32e it really broke the scheme.

The reason you're seeing very similar code generation is because the ISAs themselves are almost exactly the same and we picked the default ABIs to be efficient (ie, "long" is 64-bit on rv64i and 32-bit on rv32i).  If you use some code that is different between 32-bit and 64-bit then you can see the difference: https://godbolt.org/beta/z/Odkx2y .

After playing with this, there are some compounding issues due to the shape our ports going on here:

* Our glibc port doesn't support rv32i yet, so you shouldn't be able to include <stdint.h> on the 32-bit targets.  There appears to be something wrong with the #defines that LLVM provides and the check isn't triggering in the link.
* GCC can generate code for any target, and assuming the multilibs are installed you'll get a working setup.  You can pass arguments like `-march=rv32gc -mabi=ilp32d` to GCC 
* LLVM requires an additional `-target` argument to switch between the 32-bit and 64-bit ports, but appears to allow some amount of mixing between the outputs.  For example `-target=riscv64 -march=rv32gc` generates code, but adding the ABI in (specifically `-target=riscv64 -march=rv32gc -mabi=ilp32`) causes an error.

I hadn't really played with the LLVM things before I saw it in your beta site, but now that I know what's going on it should be easy to get it fixed.

**Comment by mattgodbolt** [comment]
Thanks for the comprehensive write-up! I had no idea it was so confusing :)

**Pull request merged**